src/ethereum_spec_tools/evm_trace.py

```diff
@@ -1,66 +1,212 @@
+        stack=[hex(i) for i in evm.stack],
+        depth=evm.message.depth + 1,
+        refund=evm.refund_counter,
+        opName=str(op).split(".")[-1],
+    )
+
+    evm.env.traces.append(new_trace)
+
+
+def capture_op_end(evm: Any) -> None:
+    """Capture the state at the end of an opcode execution."""
+    last_trace = evm.env.traces[-1]
+
+    last_trace.gasCostTraced = True
+    last_trace.errorTraced = True
+
+
+def capture_op_exception(evm: Any) -> None:
+    """Capture the state in case of exceptions."""
+    if any(evm.env.traces):
+        last_trace = evm.env.traces[-1]
+    if (
+        not any(evm.env.traces)
+        or last_trace.errorTraced
+        or last_trace.depth == evm.message.depth
+    ):
+        new_trace = Trace(
+            pc=evm.pc,
+            op="InvalidOpcode",
+            gas=hex(evm.gas_left),
+            gasCost="0x0",
+            memSize=len(evm.memory),
+            stack=[hex(i) for i in evm.stack],
+            depth=evm.message.depth + 1,
+            refund=evm.refund_counter,
+            opName="InvalidOpcode",
+            gasCostTraced=True,
+            errorTraced=True,
+            error="",
+        )
+
+        evm.env.traces.append(new_trace)
+    elif not last_trace.errorTraced:
+        last_trace.error = ""
+        last_trace.errorTraced = True
+
+
+def output_op_trace(trace: Any, json_file: Any) -> None:
+    """
+    Output a single trace to a json file.
+    """
+    dict_trace = {
+        field.name: getattr(trace, field.name)
+        for field in fields(trace)
+        if field.name not in EXCLUDE_FROM_OUTPUT
+        and getattr(trace, field.name) is not None
+    }
+
+    json.dump(dict_trace, json_file, separators=(",", ":"))
+    json_file.write("\n")
+
+
+def output_traces(
+    traces: Any, tx_hash: bytes, output_basedir: str = "."
+) -> None:
+    """
+    Output the traces to a json file.
+    """
+    tx_hash_str = "0x" + tx_hash.hex()
+    output_path = os.path.join(
+        output_basedir, f"spec-trace-{tx_hash_str}.json"
     )
-    logging.info(new_trace.custom_repr())
+    with open(output_path, "w") as json_file:
+        for trace in traces:
+
+            if getattr(trace, "precompile", False):
+                # Traces related to pre-compile are not output.
+                continue
+            output_op_trace(trace, json_file)
+
+
+def capture_evm_stop(evm: Any, op: Any) -> None:
+    """
+    Capture the state at the end of an EVM execution.
+    A stop opcode is captured.
+    """
+    if not evm.running:
+        return
+    elif len(evm.code) == 0:
+        return
+    else:
+        capture_op_start(evm, op)
+
+
+def capture_gas_and_refund(evm: Any, gas_cost: Any) -> None:
+    """
+    Capture the gas cost and refund during opcode execution.
+    """
+    if not any(evm.env.traces):
+        # In contract creation transactions, there may not be any traces
+        return
+    else:
+        last_trace = evm.env.traces[-1]
```

Comment:
This might be a bit easier to read:

```suggestion

    last_trace = evm.env.traces[-1]
```

Reply:
Done